src/System.Text.Json/tests/JsonReaderTests.cs

```diff
@@ -8,153 +7,252 @@ namespace System.Text.Json.Tests
+                        break;
                     case JsonReader.JsonTokenType.ObjectStart:
-                        json.ReadObjectStart();
-                        value = null;
+                        jsonPairs = new List<Pair>();
                         break;
                     case JsonReader.JsonTokenType.ObjectEnd:
-                        json.ReadObjectEnd();
-                        property = new Utf8String("");
-                        value = null;
+                        if (jsonPairs != null)
+                        {
+                            var jsonMembers = new Members {Pairs = jsonPairs};
+                            jsonMembersList.Add(jsonMembers);
+                            jsonObject.Members = jsonMembersList;
+                            return jsonObject;
+                        }
                         break;
                     case JsonReader.JsonTokenType.ArrayStart:
-                        json.ReadArrayStart();
-                        value = null;
                         break;
                     case JsonReader.JsonTokenType.ArrayEnd:
-                        json.ReadArrayEnd();
-                        property = new Utf8String("");
-                        value = null;
                         break;
-                    case JsonReader.JsonTokenType.PropertyName:
-                        property = json.ReadPropertyAsString();
-                        value = null;
+                    case JsonReader.JsonTokenType.Pair:
+                        var pair = new Pair
+                        {
+                            Name = (string) jsonReader.ReadString(),
+                            Value = GetValue(ref jsonReader)
+                        };
+                        if (jsonPairs != null) jsonPairs.Add(pair);
                         break;
-                    case JsonReader.JsonTokenType.PropertyValueAsString:
-                        value = json.ReadPropertyAsString();
+                    case JsonReader.JsonTokenType.Value:
                         break;
-                    case JsonReader.JsonTokenType.PropertyValueAsInt:
-                        value = json.ReadPropertyValueAsInt();
+                    case JsonReader.JsonTokenType.Finish:
                         break;
                     default:
-                        property = new Utf8String("");
-                        value = null;
-                        break;
+                        throw new ArgumentOutOfRangeException();
                 }
-                if (property != new Utf8String("") && value != null)
+            }
+            throw new FormatException("Json object was started but never ended.");
+        }
+
+        private static Array ReadArray(ref JsonReader jsonReader)
+        {
+            Array jsonArray = null;
+            List<Value> jsonValues = null;
+
+            while (jsonReader.Read())
+            {
+                switch (jsonReader.TokenType)
```

Comment:
Is this necessary? I am assuming the usage of Whitespace would be to maintain the input formatting in some way? What should the ReadWhitespace API return?